File: server/admin/src/xlsx.ts
/**
 * xlsx-выгрузка расписания DNUI -> пары в формате ассетов приложения.
 * Копия Xlsx.kt из приложения: оба разбора сверяются с одними эталонами в fixtures/
 * (xlsx.test.ts здесь, XlsxTest.kt там) - поменял правило в одном, поменяй и в другом.
 */

export interface Lesson {
  id: string;
  name: string;
  name_ru: string;
  weeks: [number, number];
  teacher: string;
  room: string;
  room_ru: string;
  building: string;
  building_ru: string;
  day: number;
  slot: number;
  start: string;
  end: string;
}
export interface Slot {
  n: number;
  start: string;
  end: string;
}
export interface Parsed {
  meta: { title: string; week1_monday: string; weeks: number };
  slots: Slot[];
  lessons: Lesson[];
}

/** Файл не похож на выгрузку расписания DNUI. Текст показывается как есть. */
export class ScheduleFormatError extends Error {}
const fail = (msg: string): never => {
  throw new ScheduleFormatError(msg);
};

// ponytail: словарь - копия RU из Xlsx.kt. Нет перевода - останется китайское название,
// его можно поправить после загрузки.
const RU: Record<string, string> = {
  "汉语口语1": "Разговорный китайский 1",
  "汉语1": "Китайский язык 1",
  "面向对象编程基础 I": "Основы ООП I",
  "前端开发技术 I": "Frontend-разработка I",
  "数据库原理与技术I": "Базы данных I",
  "机器学习 I": "Машинное обучение I",
  "金融大数据分析": "Большие данные в финансах",
  "概率论与数理统计Ⅱ": "Теорвер и матстатистика II",
  "体育3": "Физкультура 3",
};
const RU_BUILDING: Record<string, string> = { A6: "корпус A6", A7: "корпус A7", "体育馆": "спорткомплекс" };
const RU_ROOM: Record<string, string> = { "体育馆-羽毛球场": "спорткомплекс, корт для бадминтона" };

const CN_DAYS: Record<string, number> = {
  "星期一": 1, "星期二": 2, "星期三": 3, "星期四": 4, "星期五": 5, "星期六": 6, "星期日": 7,
};
const RE_WEEKS = /^(\d+)\s*-\s*(\d+)\s*周\s*(.*)$/;
const RE_ROOM = /^[A-Za-z]\d+-\S+$|^体育馆/;
const RE_PAIR = /\((\d+)-(\d+)节\)/;
const RE_TIME = /(\d{1,2}):(\d{2})\s*-\s*(\d{1,2}):(\d{2})/;
const REL_NS = "http://schemas.openxmlformats.org/officeDocument/2006/relationships";

export async function parseXlsx(data: ArrayBuffer, week1Monday: string): Promise<Parsed> {
  try {
    return await parse(await unzip(data), week1Monday);
  } catch (e) {
    if (e instanceof ScheduleFormatError) throw e;
    throw new ScheduleFormatError("Не получилось прочитать файл: " + (e as Error).message);
  }
}

interface Entry {
  name: string | null;
  from: number;
  to: number;
  teacher: string;
  room: string | null;
}

async function parse(files: Files, week1Monday: string): Promise<Parsed> {
  const sheet = await sheetOf(files);
  const cells = await cellsOf(files, sheet);
  const key = (r: number, c: number) => r * 10000 + c;

  // какие ячейки объединены по вертикали -> пара на несколько слотов
  const spans = new Map<number, number>();
  for (const m of all(sheet, "mergeCell")) {
    const [a, b] = m.getAttribute("ref")!.split(":").map(ref);
    if (a[1] === b[1] && b[0] > a[0]) spans.set(key(a[0], a[1]), b[0]);
  }

  const rows = [...cells.keys()].map((k) => Math.floor(k / 10000));
  let hdr = 0;
  for (let r = 1; r <= 10 && !hdr; r++) {
    for (const [k, v] of cells) if (Math.floor(k / 10000) === r && v in CN_DAYS) hdr = r;
  }
  if (!hdr) fail("Не нашёл строку с днями недели (星期一, 星期二…) - это точно расписание DNUI?");
  const colDay = [...cells]
    .filter(([k, v]) => Math.floor(k / 10000) === hdr && v in CN_DAYS)
    .map(([k, v]) => [k % 10000, CN_DAYS[v]] as const)
    .sort((a, b) => a[0] - b[0]);

  const slots = new Map<number, Slot>();
  const found: { e: Entry; cell: string; day: number; slot: number; start: string; end: string; span: number }[] = [];
  const last = Math.max(...rows);
  for (let r = hdr + 1; r <= last; r++) {
    const s = slotOf(cells.get(key(r, 1)) ?? "");
    if (!s) continue;
    const [n, start, end] = s;
    slots.set(n, { n, start, end });
    for (const [col, day] of colDay) {
      const text = cells.get(key(r, col));
      if (!text) continue;
      for (const e of cellOf(text)) {
        found.push({ e, cell: colName(col) + r, day, slot: n, start, end, span: (spans.get(key(r, col)) ?? r) - r + 1 });
      }
    }
  }
  if (!found.length) fail("Не распознано ни одной пары");

  for (const f of found) {
    // сдвоенная пара -> конец следующего слота
    const tail = f.span > 1 ? slots.get(f.slot + f.span - 1) : undefined;
    if (tail) f.end = tail.end;
  }
  found.sort((a, b) => a.day - b.day || a.slot - b.slot || a.e.from - b.e.from);

  const ids = new Set<string>();
  const lessons = found.map((f): Lesson => {
    const { e } = f;
    const name = e.name!;
    const room = e.room ?? fail(`Ячейка ${f.cell}: у пары «${name}» нет аудитории`);
    if (e.from < 1 || e.from > 30 || e.to < e.from || e.to > 30) {
      fail(`Ячейка ${f.cell}: странные недели ${e.from}-${e.to} у пары «${name}»`);
    }
    const id = `${f.day}-${f.slot}-${name}-${e.from}`;
    if (ids.has(id)) fail(`Ячейка ${f.cell}: пара «${name}» записана дважды`);
    ids.add(id);
    const building = room.split("-")[0];
    return {
      name,
      weeks: [e.from, e.to],
      teacher: e.teacher,
      room,
      day: f.day,
      slot: f.slot,
      start: f.start,
      end: f.end,
      building,
      name_ru: RU[name] ?? name,
      room_ru: RU_ROOM[room] ?? room,
      building_ru: RU_BUILDING[building] ?? building,
      id,
    };
  });

  return {
    meta: {
      title: cells.get(key(1, 1)) ?? "",
      week1_monday: week1Monday,
      weeks: Math.max(...found.map((f) => f.e.to)),
    },
    slots: [...slots.values()].sort((a, b) => a.n - b.n),
    lessons,
  };
}

/** Ячейка -> список пар. Формат: имя / '1-8周 препод' / [аудитория]. */
function cellOf(text: string): Entry[] {
  const out: Entry[] = [];
  let name: string | null = null;
  for (const l of text.split("\n").map((s) => s.trim()).filter(Boolean)) {
    const m = RE_WEEKS.exec(l);
    if (m) {
      out.push({ name, from: +m[1], to: +m[2], teacher: m[3].trim(), room: null });
      name = null;
    } else if (RE_ROOM.test(l)) {
      // хвостовая аудитория -> всем без своей
      for (const e of out) e.room ??= l;
    } else {
      name = l;
    }
  }
  return out.filter((e) => e.name !== null);
}

/** '第一大节\n(1-2节)\n8:00-9:40' -> [1, "08:00", "09:40"] */
function slotOf(text: string): [number, string, string] | null {
  const m = RE_PAIR.exec(text);
  const t = RE_TIME.exec(text);
  if (!m || !t) return null;
  return [Math.floor((+m[1] + 1) / 2), t[1].padStart(2, "0") + ":" + t[2], t[3].padStart(2, "0") + ":" + t[4]];
}

async function cellsOf(files: Files, sheet: Element): Promise<Map<number, string>> {
  const sst = await files.text("xl/sharedStrings.xml");
  const strings = sst
    ? all(xml(sst), "si").map((si) =>
        // rPh - фонетическая подсказка к иероглифам, в текст ячейки не входит
        all(si, "t").filter((t) => t.parentElement?.localName !== "rPh").map((t) => t.textContent).join(""),
      )
    : [];
  const out = new Map<number, string>();
  for (const c of all(sheet, "c")) {
    const [r, col] = ref(c.getAttribute("r")!);
    const v = all(c, "v")[0]?.textContent ?? null;
    const t = c.getAttribute("t");
    const text = t === "s" ? (v === null ? null : strings[+v]) : t === "inlineStr" ? all(c, "t").map((x) => x.textContent).join("") : v;
    if (text != null) out.set(r * 10000 + col, text);
  }
  return out;
}

async function sheetOf(files: Files): Promise<Element> {
  const wb = xml((await files.text("xl/workbook.xml"))!);
  const rid = all(wb, "sheet").find((s) => s.getAttribute("name") === "课表")?.getAttributeNS(REL_NS, "id")
    ?? fail("В файле нет листа «课表» - это не выгрузка расписания DNUI");
  const rels = xml((await files.text("xl/_rels/workbook.xml.rels")) ?? fail("Файл повреждён: нет workbook.xml.rels"));
  const target = all(rels, "Relationship").find((r) => r.getAttribute("Id") === rid)?.getAttribute("Target")
    ?? fail("Файл повреждён: лист 课表 не найден");
  const path = target.startsWith("/") ? target.slice(1) : "xl/" + target;
  return xml((await files.text(path)) ?? fail(`Файл повреждён: внутри нет ${path}`));
}

function xml(text: string): Element {
  const doc = new DOMParser().parseFromString(text, "application/xml");
  if (doc.getElementsByTagName("parsererror").length) fail("Файл повреждён: XML не читается");
  return doc.documentElement;
}

const all = (e: Element, tag: string) => Array.from(e.getElementsByTagNameNS("*", tag));

/** "D12" -> [12, 4] */
function ref(r: string): [number, number] {
  const letters = /^[A-Z]+/.exec(r)?.[0] ?? "";
  return [+r.slice(letters.length), [...letters].reduce((acc, ch) => acc * 26 + ch.charCodeAt(0) - 64, 0)];
}

const colName = (col: number): string =>
  col <= 26 ? String.fromCharCode(64 + col) : colName(Math.floor((col - 1) / 26)) + String.fromCharCode(65 + ((col - 1) % 26));

// ---------- zip: только чтение, без библиотек - DecompressionStream есть во всех браузерах ----------

interface Files {
  text(name: string): Promise<string | undefined>;
}

async function unzip(data: ArrayBuffer): Promise<Files> {
  const v = new DataView(data);
  // конец центрального каталога: сигнатура 0x06054b50 в последних 64 КБ
  let eocd = -1;
  for (let i = data.byteLength - 22; i >= Math.max(0, data.byteLength - 65557); i--) {
    if (v.getUint32(i, true) === 0x06054b50) {
      eocd = i;
      break;
    }
  }
  if (eocd < 0) fail("Это не xlsx-файл");
  const count = v.getUint16(eocd + 10, true);
  let p = v.getUint32(eocd + 16, true);
  const entries = new Map<string, { method: number; size: number; offset: number }>();
  const dec = new TextDecoder();
  for (let i = 0; i < count; i++) {
    if (v.getUint32(p, true) !== 0x02014b50) fail("Файл повреждён: каталог zip");
    const nameLen = v.getUint16(p + 28, true);
    const name = dec.decode(new Uint8Array(data, p + 46, nameLen));
    entries.set(name, {
      method: v.getUint16(p + 10, true),
      size: v.getUint32(p + 20, true),
      offset: v.getUint32(p + 42, true),
    });
    p += 46 + nameLen + v.getUint16(p + 30, true) + v.getUint16(p + 32, true);
  }
  if (!entries.has("xl/workbook.xml")) fail("Это не xlsx-файл");
  return {
    async text(name) {
      const e = entries.get(name);
      if (!e) return undefined;
      const start = e.offset + 30 + v.getUint16(e.offset + 26, true) + v.getUint16(e.offset + 28, true);
      const raw = new Uint8Array(data, start, e.size);
      if (e.method === 0) return dec.decode(raw);
      if (e.method !== 8) fail(`Файл повреждён: неизвестное сжатие ${e.method}`);
      const stream = new Response(raw).body!.pipeThrough(new DecompressionStream("deflate-raw"));
      return dec.decode(await new Response(stream).arrayBuffer());
    },
  };
}

// ---------- сравнение с тем, что уже на сервере ----------

const same = (a: Lesson, b: Lesson) =>
  JSON.stringify(Object.entries(a).sort()) === JSON.stringify(Object.entries(b).sort());

/**
 * Что поменяется, если заменить пары. Пропавшие id - это не только удалённые пары:
 * пара, переехавшая на другой день, тоже получает новый id, и её домашка
 * и личные заметки студентов к ней перестают показываться.
 */
export function diffLessons(before: Lesson[], after: Lesson[]) {
  const old = new Map(before.map((l) => [l.id, l]));
  const next = new Map(after.map((l) => [l.id, l]));
  return {
    added: after.filter((l) => !old.has(l.id)),
    removed: before.filter((l) => !next.has(l.id)),
    changed: after.filter((l) => old.has(l.id) && !same(old.get(l.id)!, l)),
  };
}

File: app/src/main/java/ru/valov/raspisanie/Xlsx.kt
package ru.valov.raspisanie

import org.json.JSONArray
import org.json.JSONObject
import org.w3c.dom.Element
import java.io.InputStream
import java.time.LocalDate
import java.util.TreeMap
import java.util.zip.ZipInputStream
import javax.xml.parsers.DocumentBuilderFactory

/** Пн 1-й недели семестра у встроенного расписания: в самом xlsx его нет. ПРОВЕРЬ каждый семестр. */
val WEEK1_MONDAY: LocalDate = LocalDate.of(2026, 8, 31)

/** Файл не похож на выгрузку расписания DNUI. Текст показывается пользователю как есть. */
class ScheduleFormatError(msg: String) : Exception(msg)

// ponytail: словарь в коде. Нет перевода - в приложении останется китайское название.
private val RU = mapOf(
    "汉语口语1" to "Разговорный китайский 1",
    "汉语1" to "Китайский язык 1",
    "面向对象编程基础 I" to "Основы ООП I",
    "前端开发技术 I" to "Frontend-разработка I",
    "数据库原理与技术I" to "Базы данных I",
    "机器学习 I" to "Машинное обучение I",
    "金融大数据分析" to "Большие данные в финансах",
    "概率论与数理统计Ⅱ" to "Теорвер и матстатистика II",
    "体育3" to "Физкультура 3",
)
private val RU_BUILDING = mapOf("A6" to "корпус A6", "A7" to "корпус A7", "体育馆" to "спорткомплекс")
private val RU_ROOM = mapOf("体育馆-羽毛球场" to "спорткомплекс, корт для бадминтона")

private val CN_DAYS = mapOf(
    "星期一" to 1, "星期二" to 2, "星期三" to 3, "星期四" to 4, "星期五" to 5, "星期六" to 6, "星期日" to 7,
)
private val RE_WEEKS = Regex("""(\d+)\s*-\s*(\d+)\s*周\s*(.*)""")
// ponytail: комната = "буквацифра-цифры" либо всё с 体育馆. Хватает на все 3 корпуса.
private val RE_ROOM = Regex("""^[A-Za-z]\d+-\S+$|^体育馆""")
private val RE_PAIR = Regex("""\((\d+)-(\d+)节\)""")
private val RE_TIME = Regex("""(\d{1,2}):(\d{2})\s*-\s*(\d{1,2}):(\d{2})""")
private const val REL_NS = "http://schemas.openxmlformats.org/officeDocument/2006/relationships"

/**
 * xlsx-выгрузка расписания DNUI -> JSON того же вида, что встроенные ассеты.
 * Лист 课表: столбец A - пары («第一大节 (1-2节) 8:00-9:40»), дальше дни недели;
 * в ячейке «предмет / 1-8周 препод / аудитория», объединённая по вертикали - сдвоенная пара.
 * Эталон разбора - fixtures/, с ним же сверяется парсер админки.
 */
object Xlsx {

    fun parse(input: InputStream, week1Monday: LocalDate): JSONObject =
        try {
            parse(unzip(input), week1Monday)
        } catch (e: ScheduleFormatError) {
            throw e
        } catch (e: Exception) {
            throw ScheduleFormatError("Не получилось прочитать файл: " + (e.message ?: e.javaClass.simpleName))
        }

    private class Entry(val name: String?, val from: Int, val to: Int, val teacher: String) {
        var room: String? = null
    }

    private class Found(val e: Entry, val cell: String, val day: Int, val slot: Int, val start: String, var end: String, val span: Int)

    private fun parse(files: Map<String, ByteArray>, week1Monday: LocalDate): JSONObject {
        val sheet = sheet(files)
        val cells = cells(files, sheet)

        // какие ячейки объединены по вертикали -> пара на несколько слотов
        val spans = sheet.all("mergeCell").mapNotNull {
            val (a, b) = it.getAttribute("ref").split(":").map(::ref)
            if (a.second == b.second && b.first > a.first) a to b.first else null
        }.toMap()

        val hdr = (1..10).firstOrNull { r -> cells.any { it.key.first == r && it.value in CN_DAYS } }
            ?: throw ScheduleFormatError("Не нашёл строку с днями недели (星期一, 星期二…) - это точно расписание DNUI?")
        val colDay = cells.filter { it.key.first == hdr && it.value in CN_DAYS }
            .map { it.key.second to CN_DAYS.getValue(it.value) }.toMap().toSortedMap()

        val slots = TreeMap<Int, Pair<String, String>>()
        val found = ArrayList<Found>()
        for (r in hdr + 1..cells.keys.maxOf { it.first }) {
            val (n, start, end) = slot(cells[r to 1] ?: "") ?: continue
            slots[n] = start to end
            for ((col, day) in colDay) {
                val text = cells[r to col]?.takeIf { it.isNotEmpty() } ?: continue
                for (e in cell(text)) {
                    found.add(Found(e, colName(col) + r, day, n, start, end, (spans[r to col] ?: r) - r + 1))
                }
            }
        }
        if (found.isEmpty()) throw ScheduleFormatError("Не распознано ни одной пары")

        val ids = HashSet<String>()
        val lessons = found.onEach { f ->
            // сдвоенная пара -> конец следующего слота
            if (f.span > 1) slots[f.slot + f.span - 1]?.let { f.end = it.second }
        }.sortedWith(compareBy({ it.day }, { it.slot }, { it.e.from })).map { f ->
            val e = f.e
            val name = e.name!!
            val room = e.room
                ?: throw ScheduleFormatError("Ячейка ${f.cell}: у пары «$name» нет аудитории")
            if (e.from !in 1..30 || e.to !in e.from..30)
                throw ScheduleFormatError("Ячейка ${f.cell}: странные недели ${e.from}-${e.to} у пары «$name»")
            val id = "${f.day}-${f.slot}-$name-${e.from}"
            if (!ids.add(id)) throw ScheduleFormatError("Ячейка ${f.cell}: пара «$name» записана дважды")
            val building = room.substringBefore('-')
            JSONObject()
                .put("name", name)
                .put("weeks", JSONArray().put(e.from).put(e.to))
                .put("teacher", e.teacher)
                .put("room", room)
                .put("day", f.day)
                .put("slot", f.slot)
                .put("start", f.start)
                .put("end", f.end)
                .put("building", building)
                .put("name_ru", RU[name] ?: name)
                .put("room_ru", RU_ROOM[room] ?: room)
                .put("building_ru", RU_BUILDING[building] ?: building)
                .put("id", id)
        }

        return JSONObject()
            .put("meta", JSONObject()
                .put("title", cells[1 to 1] ?: "")
                .put("week1_monday", week1Monday.toString())
                .put("weeks", found.maxOf { it.e.to }))
            .put("slots", JSONArray(slots.map { (n, t) ->
                JSONObject().put("n", n).put("start", t.first).put("end", t.second)
            }))
            .put("lessons", JSONArray(lessons))
    }

    /** Ячейка -> список пар. Формат: имя / '1-8周 препод' / [аудитория]. */
    private fun cell(text: String): List<Entry> {
        val out = ArrayList<Entry>()
        var name: String? = null
        for (l in text.split('\n').map { it.trim() }.filter { it.isNotEmpty() }) {
            val m = RE_WEEKS.matchEntire(l)
            when {
                m != null -> {
                    val (from, to, teacher) = m.destructured
                    out.add(Entry(name, from.toInt(), to.toInt(), teacher.trim()))
                    name = null
                }
                // хвостовая аудитория -> всем без своей
                RE_ROOM.containsMatchIn(l) -> out.filter { it.room == null }.forEach { it.room = l }
                else -> name = l
            }
        }
        return out.filter { it.name != null }
    }

    /** '第一大节\n(1-2节)\n8:00-9:40' -> (1, "08:00", "09:40") */
    private fun slot(text: String): Triple<Int, String, String>? {
        val m = RE_PAIR.find(text) ?: return null
        val t = RE_TIME.find(text)?.groupValues ?: return null
        return Triple(
            (m.groupValues[1].toInt() + 1) / 2,
            t[1].padStart(2, '0') + ":" + t[2],
            t[3].padStart(2, '0') + ":" + t[4],
        )
    }

    /** (строка, столбец) -> текст ячейки листа 课表. */
    private fun cells(files: Map<String, ByteArray>, sheet: Element): Map<Pair<Int, Int>, String> {
        val strings = if ("xl/sharedStrings.xml" in files) {
            xml(files, "xl/sharedStrings.xml").all("si").map { si ->
                // rPh - фонетическая подсказка к иероглифам, в текст ячейки не входит
                si.all("t").filter { it.parentNode.localName != "rPh" }.joinToString("") { it.textContent }
            }
        } else emptyList()
        val out = HashMap<Pair<Int, Int>, String>()
        for (c in sheet.all("c")) {
            val v = c.all("v").firstOrNull()?.textContent
            out[ref(c.getAttribute("r"))] = when (c.getAttribute("t")) {
                "s" -> v?.toInt()?.let { strings[it] }
                "inlineStr" -> c.all("t").joinToString("") { it.textContent }
                else -> v
            } ?: continue
        }
        return out
    }

    private fun sheet(files: Map<String, ByteArray>): Element {
        val rid = xml(files, "xl/workbook.xml").all("sheet")
            .firstOrNull { it.getAttribute("name") == "课表" }?.getAttributeNS(REL_NS, "id")
            ?: throw ScheduleFormatError("В файле нет листа «课表» - это не выгрузка расписания DNUI")
        val target = xml(files, "xl/_rels/workbook.xml.rels").all("Relationship")
            .first { it.getAttribute("Id") == rid }.getAttribute("Target")
        return xml(files, if (target.startsWith("/")) target.drop(1) else "xl/$target")
    }

    // ponytail: весь zip в память - выгрузка расписания весит килобайты.
    private fun unzip(input: InputStream): Map<String, ByteArray> {
        val out = HashMap<String, ByteArray>()
        ZipInputStream(input).use { z ->
            while (true) {
                val e = z.nextEntry ?: break
                if (!e.isDirectory) out[e.name] = z.readBytes()
            }
        }
        if ("xl/workbook.xml" !in out) throw ScheduleFormatError("Это не xlsx-файл")
        return out
    }

    private fun xml(files: Map<String, ByteArray>, path: String): Element {
        val bytes = files[path] ?: throw ScheduleFormatError("Файл повреждён: внутри нет $path")
        val f = DocumentBuilderFactory.newInstance()
        f.isNamespaceAware = true
        // DOCTYPE в xlsx не бывает, а внешние сущности из чужого файла грузить незачем.
        // Android такой настройки не знает - но и сущностей не подгружает.
        runCatching { f.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true) }
        return f.newDocumentBuilder().parse(bytes.inputStream()).documentElement
    }

    private fun Element.all(tag: String): List<Element> {
        val nl = getElementsByTagNameNS("*", tag)
        return (0 until nl.length).map { nl.item(it) as Element }
    }

    /** "D12" -> (12, 4) */
    private fun ref(r: String): Pair<Int, Int> {
        val letters = r.takeWhile { it.isLetter() }
        return r.drop(letters.length).toInt() to letters.fold(0) { acc, ch -> acc * 26 + (ch - 'A' + 1) }
    }

    private fun colName(col: Int): String =
        if (col <= 26) ('A' + col - 1).toString() else colName((col - 1) / 26) + ('A' + (col - 1) % 26)
}
